prober/derp.go

```diff
@@ -577,6 +602,246 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
+// DERP servers (from and to) and sends a test payload of a given size from one
+// to another over a TUN device at an address at the start of the usable host IP
+// range that the given tunAddress lives in. The time taken to finish the transfer
+// is recorded in `transferTime`.
+func derpProbeBandwidthTUN(ctx context.Context, transferTime *expvar.Float, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, size int64, tunAddress string) error {
+	prefix, err := netip.ParsePrefix(fmt.Sprintf("%s/30", tunAddress))
+	if err != nil {
+		return fmt.Errorf("failed to parse IP prefix: %w", err)
+	}
+
+	ipRange := netipx.RangeOfPrefix(prefix)
+	// Start of the usable host IP range from the address we have been passed in.
+	ifAddr := ipRange.From().Next()
+	// Destination address to listen on. This is the next address in the range from
+	// our ifAddr to ensure that the underlying networking stack is actually being
+	// utilized instead of being optimized away and treated as a loopback.
+	destinationAddr := ifAddr.Next()
+
+	derpProbeBandwidthTUNMu.Lock()
+	defer derpProbeBandwidthTUNMu.Unlock()
+
+	// Temporarily set up a TUN device with which to simulate a real client TCP connection
+	// tunneling over DERP. Use 1280 as our MTU as this is the minimum safe MTU used by
+	// Tailscale.
+	dev, err := tun.CreateTUN("utun9_derpprobe", 1280)
+	if err != nil {
+		return fmt.Errorf("failed to create TUN device: %w", err)
+	}
+	defer func() {
+		if err := dev.Close(); err != nil {
+			log.Printf("failed to close TUN device: %s", err)
+		}
+	}()
+	mtu, err := dev.MTU()
+	if err != nil {
+		return fmt.Errorf("failed to get TUN MTU: %w", err)
+	}
+
+	name, err := dev.Name()
+	if err != nil {
+		return fmt.Errorf("failed to get device name: %w", err)
+	}
+
+	err = configureTUN(prefix, name)
+	if err != nil {
+		return fmt.Errorf("failed to configure tun: %w", err)
+	}
+
+	// Depending on platform, we need some space for headers at the beginning
+	// of TUN packets. The below constant is more than enough space for any
+	// platform that this might run on.
+	tunStartOffset := device.MessageTransportHeaderSize
+
+	// Read packets from TUN device and forward all IPv4 packets destined to
+	// our listener's address via DERP.
+	tunReadC := make(chan error, 1)
+	go func() {
+		numBufs := wgconn.IdealBatchSize
+		bufs := make([][]byte, 0, numBufs)
+		sizes := make([]int, numBufs)
+		for range numBufs {
+			bufs = append(bufs, make([]byte, mtu+tunStartOffset))
+		}
+
+		src, dst := make([]byte, 4), make([]byte, 4)
+		for {
+			n, err := dev.Read(bufs, sizes, tunStartOffset)
+			if err != nil {
+				return
+			}
+
+			for i := range n {
+				pkt := bufs[i][tunStartOffset : sizes[i]+tunStartOffset]
+				// Skip everything except valid IPv4 packets
+				if len(pkt) < 20 {
+					// Doesn't even have a full IPv4 header
+					continue
+				}
+				if pkt[0]>>4 != 4 {
+					// Not IPv4
+					continue
+				}
+
+				destinationAddrBytes := destinationAddr.As4()
```

Comment:
we don't need to do this for every packet, we can do it outside both containing for loops

Reply:
also, the comparison below is slice-based, `.AsSlice()`?